From: Roland McGrath <roland@redhat.com>
Date: Wed, 9 Apr 2008 08:30:06 +0000 (-0700)
Subject: x86 vDSO: don't map 32-bit vdso when disabled
X-Git-Tag: v2.6.26-rc1~1154^2~608
X-Git-Url: http://git.kernel.org/?p=linux%2Fkernel%2Fgit%2Ftorvalds%2Flinux-2.6.git;a=commitdiff_plain;h=5de253cc5b1f565f7aeb5bacd67bac37e943ceef

x86 vDSO: don't map 32-bit vdso when disabled

We map a VMA for the 32-bit vDSO even when it's disabled, which is stupid.
For the 32-bit kernel it's the vdso_enabled boot parameter/sysctl
and for the 64-bit kernel it's the vdso32 boot parameter/syscall32 sysctl.

When it's disabled, we don't pass AT_SYSINFO_EHDR so processes don't use
the vDSO for anything, but we still map it.  For the non-compat vDSO,
this means we're always putting an extra VMA somewhere, maybe lousing
up the control of the address space the user was hoping for.

Honor the setting by doing nothing in arch_setup_additional_pages.

[ also see: "x86 vDSO: don't use disabled vDSO for signal trampoline" ]

Signed-off-by: Roland McGrath <roland@redhat.com>
Signed-off-by: Ingo Molnar <mingo@elte.hu>
---

Index: 2008-devel/arch/x86/vdso/vdso32-setup.c
===================================================================
--- 2008-devel.orig/arch/x86/vdso/vdso32-setup.c
+++ 2008-devel/arch/x86/vdso/vdso32-setup.c
@@ -325,6 +325,9 @@ int arch_setup_additional_pages(struct l
 	int ret = 0;
 	bool compat;
 
+	if (vdso_enabled == VDSO_DISABLED)
+		return 0;
+
 	down_write(&mm->mmap_sem);
 
 	/* Test compat mode once here, in case someone
